storage/rocksdb/clone/client.cc

```diff
@@ -0,0 +1,866 @@
+        mysql_file_create(myrocks::rdb_clone_client_file_key, path.c_str(),
+                          m_clone_file_mode, create_flags, MYF(MY_WME));
+    if (fd < 0) {
+      mysql_rwlock_unlock(&m_files_lock);
+      return save_and_handle_error(thd, ER_CANT_CREATE_FILE, path);
+    }
+
+#ifdef __APPLE__
+    if (use_direct_io && fcntl(fd, F_NOCACHE, 1) == -1) {
+      mysql_rwlock_unlock(&m_files_lock);
+      mysql_file_close(fd, MYF(MY_WME));
+      return myrocks::clone::return_error(
+          ER_INTERNAL_ERROR, "Failed to set fcntl(F_NOCACHE) on a clone file");
+    }
+#endif
+
+    const Ha_clone_file clone_file{
+#ifndef __APPLE__
+        use_direct_io,
+#else
+        false,
+#endif
+        Ha_clone_file::FILE_DESC,
+        {fd}};
+    // We could check file names for duplicates, but doing that efficiently
+    // would require another data structure.
+    const auto insert_result = m_files_in_progress.emplace(
+        std::piecewise_construct, std::forward_as_tuple(new_file.get_id()),
+        std::forward_as_tuple(clone_file, new_file.get_size(),
+                              std::move(name)));
+    mysql_rwlock_unlock(&m_files_lock);
+
+    if (!insert_result.second) {
+      mysql_file_close(clone_file.file_desc, MYF(MY_WME));
+      return myrocks::clone::return_error(
+          ER_CLONE_PROTOCOL, "Incorrect file metadata received from donor");
+    }
+  }
+
+  return HA_EXIT_SUCCESS;
+}
+
+int client::apply_chunk(client_chunk_metadata &&chunk, Ha_clone_cbk &cbk,
+                        const THD *thd) {
+  mysql_rwlock_rdlock(&m_files_lock);
+  const auto itr = m_files_in_progress.find(chunk.get_file_id());
+  if (itr == m_files_in_progress.end()) {
+    mysql_rwlock_unlock(&m_files_lock);
+    return myrocks::clone::return_error(
+        ER_CLONE_PROTOCOL, "Incorrect chunk metadata received from donor");
+  }
+  auto &file_in_progress = itr->second;
+  mysql_rwlock_unlock(&m_files_lock);
+
+  const auto file_handle = file_in_progress.get_file();
+
+  auto err = cbk.apply_file_cbk(file_handle);
+
+  DBUG_EXECUTE_IF("myrocks_clone_apply_fail", {
+    err = ER_ERROR_ON_WRITE;
+    set_my_errno(ENOSPC);
+  });
+
+  if (err != 0)
+    return save_and_handle_error(thd, err, file_in_progress.get_name());
+
+  if (file_in_progress.advance_pos(chunk.get_size())) {
+    err = mysql_file_close(file_in_progress.get_file().file_desc, MYF(MY_WME));
+    if (err != 0)
+      return save_and_handle_error(thd, err, file_in_progress.get_name());
+
+    mysql_rwlock_wrlock(&m_files_lock);
+    m_completed_files.emplace_front(chunk.get_file_id(),
+                                    file_in_progress.get_applied_offset());
+    ++m_completed_files_count;
+    const auto erase_result = m_files_in_progress.erase(chunk.get_file_id());
+    mysql_rwlock_unlock(&m_files_lock);
+
+    if (erase_result != 1) {
+      return myrocks::clone::return_error(
+          ER_CLONE_PROTOCOL, "Incorrect chunk metadata received from donor");
```

Comment:
Same as above - adding additional file metadata information, and it might be helpful to have a different message too.

Reply:
Done in https://github.com/facebook/mysql-5.6/pull/1250/commits/fd516a5bacc9e89510425c1ecee4266edaeb2e08 (later will squash)